earth2grid/_regrid.py

```diff
@@ -48,22 +50,216 @@ def forward(self, x):
+        )
+
+        stride = x_coords.size(-1)
+        index = torch.stack(
+            [
+                x_l_idx + stride * y_l_idx,
+                x_u_idx + stride * y_l_idx,
+                x_l_idx + stride * y_u_idx,
+                x_u_idx + stride * y_u_idx,
+            ],
+            dim=-1,
+        )
+        self.register_buffer("weights", weights)
+        self.register_buffer("mask", mask)
+        self.register_buffer("index", index)
+
+    def forward(self, z: torch.Tensor):
+        """
+        Interpolate the field
+
+        Args:
+            z: shape [Y, X]
+        """
+        *shape, y, x = z.shape
+        zrs = z.view(-1, y * x).T
+        # using embedding bag is 2x faster on cpu and 4x on gpu.
+        output = torch.nn.functional.embedding_bag(self.index, zrs, per_sample_weights=self.weights, mode='sum')
+        interpolated = torch.full(
+            [self.mask.numel(), zrs.shape[1]], fill_value=self.fill_value, dtype=z.dtype, device=z.device
+        )
+        interpolated.masked_scatter_(self.mask.unsqueeze(-1), output)
+        interpolated = interpolated.T.view(*shape, self.mask.numel())
+        return interpolated
+
+
+class S2NearestNeighborInterpolator(torch.nn.Module):
+    """Bilinear interpolation for a non-uniform grid"""
+
+    def __init__(
+        self,
+        src_lon: torch.Tensor,
+        src_lat: torch.Tensor,
+        dest_lon: torch.Tensor,
+        dest_lat: torch.Tensor,
+    ) -> None:
+        """
+
+        Args:
+            src_lon: (m,) source longitude in degrees E
+            src_lat: (m,) source latitude in degrees N
+            dest_lon: (n,) output longitude in degrees E
+            dest_lat: (n,) output latitude in degrees N
+
+        """
+        super().__init__()
+        src_lon = torch.deg2rad(src_lon.cpu())
+        src_lat = torch.deg2rad(src_lat.cpu())
+
+        dest_lon = torch.deg2rad(dest_lon.cpu())
+        dest_lat = torch.deg2rad(dest_lat.cpu())
+
+        vec = torch.stack(ang2vec(src_lon, src_lat), -1)
+
+        # havesign distance and euclidean are monotone for points on S2 so can use 3d lookups.
+        self.tree = spatial.KDTree(vec)
+        vec = torch.stack(ang2vec(dest_lon.cpu(), dest_lat.cpu()), -1)
+        _, neighbors = self.tree.query(vec, k=1)
+        self.register_buffer("index", torch.as_tensor(neighbors).view(-1, 1))
+
+    def forward(self, z: torch.Tensor):
+        """
+        Interpolate the field
+
+        Args:
+            z: shape [*, m]
+
+        Returns:
+            shape [*, n]
+        """
+        *shape, x = z.shape
+        zrs = z.view(-1, x).T
+        # using embedding bag is 2x faster on cpu and 4x on gpu.
+        output = torch.nn.functional.embedding_bag(self.index, zrs, mode='sum')
+        output = output.T.view(*shape, -1)
+        return output
+
+
+class S2LinearBarycentricInterpolator(torch.nn.Module):
```

Comment:
Would be good to add a test that saves a checkpoint of an initialized instance of this and loads that back in. Would be very useful to avoid repeated calls to `spatial.KDTree` especially in ETL jobs where a regridding script might be invoked multiple times on different inputs.

Reply:
I'll make an issue for this. I think it may require some thought and potentially backwards breaking changes to apply this consistently across the suite of re-gridders.

I'm not sure the most elegant way to allow ser/deser of "buffer-only" modules like this. Do you know of an example for this e.g. with normalization.

Reply:
I think we don't necessarily need a proper serialization/deserialization pipeline here right? It might be sufficient to create a classmethod that takes in a state dict and returns an initialized instance from that?

Reply:
I refactored this so that the `S2NearestNeighborInterpolator` now returns a serializeable `Regridder` object. Let me know what you think.